luaui/Widgets/gui_spectator_hud.lua

```diff
@@ -0,0 +1,2307 @@
+
+    local function isReclaimerUnit(unitDefID, unitDef)
+        return unitDef.isBuilder and not unitDef.isFactory
+    end
+
+    local function isEnergyConverter(unitDefID, unitDef)
+        return unitDef.customParams.energyconv_capacity and unitDef.customParams.energyconv_efficiency
+    end
+
+    local function isBuildPower(unitDefID, unitDef)
+        return unitDef.buildSpeed and (unitDef.buildSpeed > 0)
+    end
+
+    local function isArmyUnit(unitDefID, unitDef)
+        -- anything with a least one weapon and speed above zero is considered an army unit
+        return unitDef.weapons and (#unitDef.weapons > 0) and unitDef.speed and (unitDef.speed > 0)
+    end
+
+    local function isDefenseUnit(unitDefID, unitDef)
+        return unitDef.weapons and (#unitDef.weapons > 0) and (not unitDef.speed or (unitDef.speed == 0))
+    end
+
+    local function isUtilityUnit(unitDefID, unitDef)
+        return unitDef.customParams.unitgroup == 'util'
+    end
+
+    local function isEconomyBuilding(unitDefID, unitDef)
+        return (unitDef.customParams.unitgroup == 'metal') or (unitDef.customParams.unitgroup == 'energy')
+    end
+
+    unitDefsToTrack = {}
+    unitDefsToTrack.commanderUnitDefs = {}
+    unitDefsToTrack.reclaimerUnitDefs = {}
+    unitDefsToTrack.energyConverterDefs = {}
+    unitDefsToTrack.buildPowerDefs = {}
+    unitDefsToTrack.armyUnitDefs = {}
+    unitDefsToTrack.defenseUnitDefs = {}
+    unitDefsToTrack.utilityUnitDefs = {}
+    unitDefsToTrack.economyBuildingDefs = {}
+
+    for unitDefID, unitDef in ipairs(UnitDefs) do
+        if isCommander(unitDefID, unitDef) then
+            unitDefsToTrack.commanderUnitDefs[unitDefID] = true
+        end
+        if isReclaimerUnit(unitDefID, unitDef) then
+            unitDefsToTrack.reclaimerUnitDefs[unitDefID] = { unitDef.metalMake, unitDef.energyMake }
+        end
+        if isEnergyConverter(unitDefID, unitDef) then
+            unitDefsToTrack.energyConverterDefs[unitDefID] = tonumber(unitDef.customParams.energyconv_capacity)
+        end
+        if isBuildPower(unitDefID, unitDef) then
+            unitDefsToTrack.buildPowerDefs[unitDefID] = unitDef.buildSpeed
+        end
+        if isArmyUnit(unitDefID, unitDef) then
+            unitDefsToTrack.armyUnitDefs[unitDefID] = { unitDef.metalCost, unitDef.energyCost }
+        end
+        if isDefenseUnit(unitDefID, unitDef) then
+            unitDefsToTrack.defenseUnitDefs[unitDefID] = { unitDef.metalCost, unitDef.energyCost }
+        end
+        if isUtilityUnit(unitDefID, unitDef) then
+            unitDefsToTrack.utilityUnitDefs[unitDefID] = { unitDef.metalCost, unitDef.energyCost }
+        end
+        if isEconomyBuilding(unitDefID, unitDef) then
+            unitDefsToTrack.economyBuildingDefs[unitDefID] = { unitDef.metalCost, unitDef.energyCost }
+        end
+    end
+end
+
+local function addToUnitCache(teamID, unitID, unitDefID)
+    local function addToUnitCacheInternal(cache, teamID, unitID, value)
```

Comment:
Fixed in ff3c36a72

nvm this didn't work. Reverted in 5525cc909. Without the revert widget fails:

```
Error in Initialize(): [string "LuaUI/Widgets/gui_spectator_hud.lua"]:453: attempt to index local 'value' (a number value)
```

Reply:
You'll also need to update any locations where it's called to remove those parameters as well. The error you're getting is because it expects the second argument to be `value`, but it's getting `teamID`, so all of the values end up set to the `teamID` instead.